dask/local.py

```diff
@@ -405,15 +405,16 @@ def get_async(apply_async, num_workers, dsk, result, cache=None,
         started_cbs = []
         succeeded = False
         try:
+            keyorder = order(dsk)
+
+            state = start_state_from_dask(dsk, cache=cache, sortkey=keyorder.get)
```

Comment:
Sorry for the late review. This change breaks caching because the start callback needs to run *before* we initialize state, otherwise we don't replace tasks with computed keys properly, resulting in an invalid `state` (we'll compute dependencies for cached task `x`, then the callback will replace task `x` with its computed value, invalidating the computed dependencies).